tools/matedit/src/main.cpp

```diff
@@ -0,0 +1,170 @@
+        "\n"
+        "       This command will remove the text-based shaders when writing the output material file.\n"
+        "\n"
+        "       If script exits with a non-zero exit code, MATEDIT will terminate with error. Multiple\n"
+        "       invocations of script may be launched in parallel.\n"
+        "\n"
+        "Example:\n"
+        "   MATEDIT -o out.cmat -i in.cmat --type metal external-compile -- ./my_compile-script.sh --sdk iphones\n"
+    );
+
+    const std::string from("MATEDIT");
+    for (size_t pos = usage.find(from); pos != std::string::npos; pos = usage.find(from, pos)) {
+        usage.replace(pos, from.length(), execName);
+    }
+    printf("%s", usage.c_str());
+}
+
+static int handleArguments(int argc, char* argv[], Config* config) {
+    static constexpr const char* OPTSTR = "hi:o:t:";
+    static const struct option OPTIONS[] = {
+            { "help",               no_argument,       nullptr, 'h' },
+            { "input",              required_argument, nullptr, 'i' },
+            { "output",             required_argument, nullptr, 'o' },
+            { "type",               required_argument, nullptr, 't' },
+            { nullptr, 0, nullptr, 0 }  // termination of the option list
+    };
+
+    int opt;
+    int optionIndex = 0;
+
+    while ((opt = getopt_long(argc, argv, OPTSTR, OPTIONS, &optionIndex)) >= 0) {
+        std::string arg(optarg ? optarg : "");
+        switch (opt) {
+            default:
+            case 'h':
+                printUsage(argv[0]);
+                exit(0);
+            case 'i':
+                config->inputFile = arg;
+                break;
+            case 'o':
+                config->outputFile = arg;
+                break;
+            case 't':
+                if (arg == "metal") {
+                    // no-op for now
+                } else {
+                    std::cerr << "Unrecognized shader type flag: '" << arg << "'. Must be 'metal'"
+                              << std::endl;
+                    exit(1);
+                }
+        }
+    }
+
+    return optind;
+}
+
+int main(int argc, char* argv[]) {
+    Config config;
+    int optionIndex = handleArguments(argc, argv, &config);
+
+    if (config.inputFile.isEmpty() || config.outputFile.isEmpty()) {
+        printUsage(argv[0]);
+        return 1;
+    }
+
+    if (!config.inputFile.exists()) {
+        std::cerr << "The source material " << config.inputFile << " does not exist." << std::endl;
+        return 1;
+    }
+
+    int numArgs = argc - optionIndex;
+    if (numArgs < 1) {
+        printUsage(argv[0]);
+        return 1;
+    }
+
+    const std::string command = argv[optionIndex++];
+    if (command != "external-compile") {
+        std::cerr << "Unrecognized command: '" << command << "'. Must be 'external-compile'"
+                  << std::endl;
+        return 1;
+    }
+
+    for (int i = optionIndex; i < argc; ++i) {
+        if (strcmp(argv[i], "--") == 0) {
```

Comment:
nit: with this, technically you could have
```external-compile my-script.sh -- -my-script-option=1```
which I don't know is expected to be valid.

Reply:
Hmm, good point. Maybe we should only ignore "--" args between "external-compile" and the path to the script.